File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = -1
    stk = [(k, [], 0)]
    
    while stk:
        curr_k, visited, dist = stk.pop()
        answer = max(answer, dist)
        
        for i in range(len(dungeons)):
            if i not in visited and curr_k >= dungeons[i][0]:
                stk.append((curr_k-dungeons[i][1],visited + [i], dist+1))
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>

using namespace std;
vector<bool> visited;
int answer = 0;

void dfs(int k,vector<vector<int>> dungeons, int cnt){
    answer = max(answer, cnt);
    
    for(int i=0; i<dungeons.size(); i++){
        if(!visited[i] && dungeons[i][0]<=k){
            visited[i]=true;
            dfs(k-dungeons[i][1], dungeons, cnt+1);
            visited[i]=false;
        }
    }
    
}


int solution(int k, vector<vector<int>> dungeons) {
    visited.assign(dungeons.size(), false);
    dfs(k, dungeons, 0);
    return answer;
}